jax/_src/pallas/primitives.py

```diff
@@ -956,3 +958,712 @@ def _lower_fun(*lower_fun_args):
     return out[:num_return_values]
 
   return mlir.lower_fun(_lower_fun, multiple_results=True)(ctx, *args)
+
+
+class DeviceIdType(enum.Enum):
+  MESH = "mesh"
+  LOGICAL = "logical"
+
+
+def check_sem_avals(
+    sem_aval, sem_transforms_avals, name, allowed_semaphore_types=None
+):
+  if allowed_semaphore_types is None:
+    allowed_semaphore_types = {
+        pallas_core.semaphore,
+        pallas_core.barrier_semaphore,
+        # For interpret mode.
+        pallas_core.SEMAPHORE_INTERPRET_DTYPE,
+    }
+  if not isinstance(sem_aval, state.AbstractRef):
+    raise ValueError(f"Cannot {name} on a non-semaphore Ref: {sem_aval}")
+  sem_shape = sem_aval.shape
+  if sem_transforms_avals:
+    sem_shape = sem_transforms_avals[-1].get_indexer_shape()
+  if sem_shape:
+    raise ValueError(f"Cannot {name} on a non-()-shaped semaphore: {sem_shape}")
+  sem_dtype = sem_aval.dtype
+  # Uncomment check below when semaphores on GPUs get allocated in scratch spaces
```

Comment:
I don't follow this comment. What does a dtype check have to do with scratch spaces?

Reply:
It refers to the issue I have with declaring semaphores with Mosaic-GPUs:

In this set of PRs, the semaphores are allocated in user-code with `jnp.array` and are explicitly passed in to the Pallas kernel. For a better UX, it would be better that the user declares the semaphores he needs in the scratch_space arg (what Mosaic does) and Mosaic-GPU will allocate and pass the semaphores under the cover to the kernel.
So right now with jnp.int32 dtype it would fail here for Mosaic-GPU path.

Reply:
One thing to consider here is that we'll need to make sure this parameter is aliased to a kernel output in HLO or else you'll be violating HLO assumptions (that you don't modify inputs, unless they are also aliased as an output).

Reply:
yes correct they need to be used with input_output_aliasing.

Reply:
@apaszke I getting `TypeError: Value '|V0' with dtype |V0 is not a valid JAX array type. Only arrays of numeric types are supported by JAX.` using `bitcast`.

Reply:
Can you share a more complete stack trace? I can't tell what's wrong just from the error message :/